contracts/Controller.sol

```diff
@@ -230,42 +235,63 @@ contract Controller is ReentrancyGuard, Ownable {
      * @dev For each action in the action Array, run the corresponding action
      * @param _actions An array of type Actions.ActionArgs[] which expresses which actions the user want to execute.
      */
-    function _runActions(Actions.ActionArgs[] memory _actions) internal {
+    function _runActions(Actions.ActionArgs[] memory _actions) internal returns (MarginAccount.Vault memory) {
+        MarginAccount.Vault memory vault;
+
+        uint256 prevActionVaultId;
+        bool isActionVaultStored;
+
         for (uint256 i = 0; i < _actions.length; i++) {
             Actions.ActionArgs memory action = _actions[i];
             Actions.ActionType actionType = action.actionType;
 
-            uint256 prevActionVaultId;
-
             if (actionType == Actions.ActionType.OpenVault) {
                 // check if this action is manipulating the same vault as all other actions, other than SettleVault
-                require(
-                    prevActionVaultId == 0 || action.vaultId == prevActionVaultId,
-                    "Controller: can not run actions on different vaults"
+                (prevActionVaultId, isActionVaultStored) = _checkActionsVaults(
+                    prevActionVaultId,
+                    action.vaultId,
+                    isActionVaultStored
                 );
-                prevActionVaultId = action.vaultId;
 
                 _openVault(Actions._parseOpenVaultArgs(action));
             }
         }
+
+        return vault;
     }
 
     /**
      * @notice verify vault final state after executing all actions
-     * @param _accountOwner account owner address
-     * @param _vaultId vault id related to the vault that will be checked
+     * @param _vault final vault state
      */
-    function _verifyFinalState(address _accountOwner, uint256 _vaultId) internal view {
-        MarginAccount.Vault memory vault = vaults[_accountOwner][_vaultId];
-
-        if (vault.shortOtokens.length > 0) {
+    function _verifyFinalState(MarginAccount.Vault memory _vault) internal view {
+        if (_vault.shortOtokens.length > 0) {
```

Comment:
is there an issue with verifying final state if no short tokens in a vault?

Reply:
yeah, `vault.shortOtokens[0]` will revert with invalid opcode error, because w ebasically accessing index 0 which is an invalid index.

Valid indexes in any array are between 0 and array.length - 1.